# Assignment 1
The assignment is accompanied by a setup notebook that contains boilerplate code for each exercise. Some of the answers depend on a particular dataset split or seed, so consult the boilerplate code for specifics even if you would like to program everything from scratch.

## Software Libraries

For this assignment, we strongly encourage you to use NumPy for your implementations. While scikit-learn provides convenient implementations of many algorithms, we recommend implementing the core algorithms from scratch using NumPy for several important reasons:

**Educational Value**: Implementing algorithms from scratch deepens your understanding of the underlying mathematics and computational principles. You'll gain insight into the inner workings that you might miss when using high-level libraries.

**Control and Transparency**: NumPy gives you precise control over your implementations without the hidden complexity that can exist in scikit-learn's optimized code. This transparency is crucial for debugging and understanding exactly what your code is doing.

**Avoiding Implementation Pitfalls**: If you do choose to use scikit-learn, please be extremely careful. The library contains many implementation details, optimizations, and default parameters that can significantly affect your results. These details are not always obvious from the documentation and can lead to incorrect conclusions if not properly understood.

If you decide to use scikit-learn despite this recommendation, please thoroughly read the documentation and understand all parameters and their effects.

## A Note On Using Language Models
As a student in the age of AI, you have without a doubt used language models to study course material, answer questions or even solve assignments for you. Using other sources than the course material to learn is valuable indeed, but a word of caution:

The point of these exercises is for you to develop as an engineer and as a scientist. Using language models to directly answer the question might get you through the assignment, but it is liable to damage your development in the long run. In the end **someone needs to understand what the model is talking about in order to validate the outcomes.**

It is quite likely you will not end up designing gradient descent algorithms for a living. But do not confuse the specifics of this problem set with the broader educational value of **solving a difficult problem yourself from start to finish**.

Education is about more than gaining skills. It is also about a work ethic, a way of approaching problems, scientific rigor, managing your thought process, using the tools at your disposal and recognizing where your strengths and weaknesses lie. Using language models heedlessly runs the risk of depriving you of the opportunity to struggle through that process.

We can only ask you to adopt more responsibility for your education yourself. Consider not denying yourself the challenge and pressure necessary for growth, and instead use the tools you have to deepen your understanding. In the end, the responsibility is yours.

# 1. Linear Algebra for Datamining and Machine Learning

This assignment focuses on linear algebra concepts essential for understanding 
Principal Component Analysis (PCA), covariance analysis, and feature transformations.

Let $X \in \mathbb{R}^{m\times p}$ be a centered data matrix where each row 
represents an observation and each column represents a feature. The centering 
condition means each column has zero mean: $\sum_{i=1}^m X_{i,j} = 0$ for all $j$.

## 1a. Eigenvalues and Total Variance
Given the eigendecomposition of the sample covariance matrix
$$
C = \frac{1}{m-1} X^\top X = Q \Lambda Q^\top
$$
where $Q$ contains the eigenvectors and $\Lambda$ is diagonal with eigenvalues 
$\lambda_1, \lambda_2, \ldots, \lambda_p$, the total variance in the dataset equals:
$$
\text{Total Sample Variance} = \operatorname{tr}(C) = \frac{1}{m-1}\operatorname{tr}(X^\top X) = \lambda_{max}
$$

○ False  
○ True

## 1b. Covariance for Centered Data
For a centered data matrix $ X $, the sample covariance between features $ j $ and $ k $ can be computed as:

$$
\operatorname{Cov}(j,k) = \frac{1}{m-1} \, X_{(:,j)}^\top X_{(:,k)}
$$

where $ X_{(:,j)} $ denotes the $ j $-th column of $ X $.

○ False  
○ True

## 1c. Covariance Under Linear Transformations

If we apply a linear transformation

$$
Y = X W
$$

where $ W \in \mathbb{R}^{p\times q} $ is a transformation matrix, then the covariance matrix of the transformed features is:

$$
C_Y = W^\top C_X W
$$

where

$$
C_X = \frac{X^\top X}{m-1}
$$

is the covariance matrix of $ X $.

○ False  
○ True

## 1d. Matrix Norms and Trace Properties
For any matrix $ A \in \mathbb{R}^{n\times d} $, the following relationship holds between row-wise and column-wise squared norms:

$$
\operatorname{tr}(A A^\top) = \sum_{i=1}^n \|A_{(i,:)}\|^2 = \sum_{j=1}^d \|A_{(:,j)}\|^2
$$

○ False  
○ True


# 2. Data Normalization

In this exercise, you'll implement and explore different data normalization techniques, fundamental preprocessing steps in machine learning and data analysis. Raw data often comes with features on vastly different scales. Imagine comparing house prices in dollars with number of bedrooms, or pixel intensities (0-255) with age in years. Without proper scaling, algorithms can be biased toward features with larger numerical ranges, leading to poor performance.

Data normalization addresses this by transforming features to comparable scales while preserving the underlying relationships in the data. Different scaling methods make different assumptions about the data distribution and have varying robustness to outliers:

- **Min-Max Scaling** transforms features to a fixed range (typically [0,1]), preserving the original distribution shape but being sensitive to outliers
- **Standardization (Z-score)** centers data around zero with unit variance, assuming normally distributed data
- **Robust Scaling** uses median and interquartile range, making it less sensitive to extreme values

We will work with a small product sales dataset that contains information about five products.
Each row corresponds to a product (ID 0–4), and each column is a feature:

- *F1*: UnitsSold → Number of units sold
- *F2*: Revenue → Total revenue generated (in euros)
- *F3*: Returns → Number of returned items

The dataset is given by:
```
ID  F1   F2     F3
0   50   2000   5 
1   70   2500   7 
2   65   2100   6 
3   500  20000  40
4   60   2200   5 
```

This dataset is deliberately chosen because the scales are very different:
- UnitsSold ranges from 50 to 500 (with an outlier at 500).
- Revenue ranges from 2000 to 20000, much larger values than the others.
- Returns are small numbers (5–40).

You will implement and apply different scaling strategies to understand how each transformation affects the data distribution and individual values.

You are asked to:

- Implement a function that applies **min-max scaling** to transform features to the range [0,1]:
  ```python
  def minmax_scale(data: list) -> list:
  ```
  using the transformation:
  $$
  x_{scaled} = \frac{x - x_{min}}{x_{max} - x_{min}}
  $$

- Implement a function that applies **standardization** (z-score normalization):
  ```python
  def standardize(data: list) -> list:
  ```
  using the transformation:
  $$
  x_{standardized} = \frac{x - \mu}{\sigma}
  $$
  where $\mu$ is the mean and $\sigma$ is the standard deviation.

- Implement a function that applies **robust scaling**:
  ```python
  def robust_scale(data: list) -> list:
  ```
  using the transformation:
  $$
  x_{robust} = \frac{x - \text{median}(x)}{\text{IQR}(x)}
  $$
  where IQR is the interquartile range (Q3 - Q1).

**All final values should be rounded to two decimal places.**

For each scaling method, you should:
- Apply the transformation to the specified feature column
- Report the transformed values
- Analyze how the scaling affects the data distribution

## 2a. Min-Max Scaling

Apply min-max scaling to feature **F1** and report the complete transformed column:

**Processed values for F1 with min-max scaling:**
$$
F1_{minmax} = [?, ?, ?, ?, ?]
$$

## 2b. Standardization

Apply standardization to feature **F2** and report the specific transformed value for ID 4:

**Processed value for ID 4's F2 after standardization:**
$$
F2_{standardized}[ID=4] = ?
$$

## 2c. Robust Scaling

Apply robust scaling to feature **F3** and report the specific transformed value for ID 3:

**Processed value for ID 3's F3 after robust scaling:**
$$
F3_{robust}[ID=3] = ?
$$

In [2]:
import numpy as np
import pandas as pd
dataset = pd.DataFrame({'F1': [50, 70, 65, 500, 60], 'F2': [2000, 2500, 2100, 20000, 2200], 'F3': [5, 7, 6, 40, 5]})
dataset.head(10)

# 2a. Min-Max Scaling F1
def min_max_scale(data: list) -> list:
    x_min = min(data)
    x_max = max(data)
    x_scaled = [round((x-x_min)/(x_max-x_min), 2) for x in data]
    return x_scaled

dataset['F1_MinMax'] = min_max_scale(dataset['F1'].tolist())

# 2b. Standardization F2
def standardize(data: list) -> list:
    mean = np.mean(data)
    sd = np.std(data)
    x_standardized = [round((x-mean)/sd, 2) for x in data]
    return x_standardized
dataset['F2_Standardized'] = standardize(dataset['F2'].tolist())
ID_4 = dataset['F2_Standardized'][4]
print(f'ID_4: {ID_4}')

# 2c. Robust Scaling F3
def robust_scale(data: list) -> list:
    median = np.median(data)
    iqr = np.percentile(data, 75) - np.percentile(data, 25)
    x_robust = [round((x-median)/iqr, 2) for x in data]
    return x_robust
dataset['F3_Robust'] = robust_scale(dataset['F3'].tolist())
ID_3 = dataset['F3_Robust'][3]
print(f'ID_3: {ID_3}')

dataset.head(10)

ID_4: -0.5
ID_3: 17.0


F1     F2  F3  F1_MinMax  F2_Standardized  F3_Robust
0   50   2000   5       0.00            -0.53       -0.5
1   70   2500   7       0.04            -0.46        0.5
2   65   2100   6       0.03            -0.51        0.0
3  500  20000  40       1.00             2.00       17.0
4   60   2200   5       0.02            -0.50       -0.5

# 3. Gradient Descent

In this exercise, you'll implement and explore gradient descent, a cornerstone of modern optimization. The method is simple in idea but powerful in application: we iteratively update our variables in the direction that most rapidly decreases the function, based on the **negative gradient**. While coordinate descent updates one variable at a time, gradient descent moves in the direction of steepest descent across all coordinates simultaneously. This often leads to:

- Faster convergence when variables are tightly coupled or when the function isn’t naturally separable,
- More direct progress toward a minimum, especially in smooth, well-behaved functions,
- Better behavior in high-dimensional settings where dependencies across variables are strong.

You will be exploring the concept by taking a look at minimizing **Himmelblau’s function**, a classical example in optimization with multiple local minima. It is often used to study the behavior of optimization algorithms such as gradient descent in non-convex settings.

The function is given by
$$
f(u, v) = (u^2 + v - 11)^2 + (u + v^2 - 7)^2
$$
where $u, v \in \mathbb{R}$.

You will use the **gradient descent** algorithm to find a minimum of $f$. Specifically, you will investigate how different **step-size strategies** and **initial points** affect convergence.

You are asked to:

- Implement a function that takes a point $(u, v)$ and returns the gradient $\nabla f(u, v)$ at that point.
- Implement a function
  ```python
  def gradient_descent(f, grad_f, eta, u0, v0, max_iter=100) -> tuple[list, list]:
  ```
  that performs the update rule:
  $$
  x_{t+1} \leftarrow x_t - \eta(t) \nabla f(x_t)
  $$
  where input $x_t$ is given by `x_t = (u_t, v_t)` and `eta(t)` defines the python method that returns the step size at iteration $t$. It is useful to make it return both the path and the computed values at each step.
- Using this setup, run 100 steps of gradient descent starting at $(u_0, v_0) = (4, -5)$ and evaluate different step-size strategies.
- Evaluate different starting points.

For each of the following strategies, report:

- The final function value:
  $$
  f(u_{100}, v_{100}) =
  $$
- The best (lowest) value reached during training:
  $$
  \min_{1 \leq t \leq 100} f(u_t, v_t) =
  $$

## 3a. Constant Step Size

Implement a constant step-size strategy $\eta=c$:
```python
def eta_const(t,c=1e-3) -> float:
```

## 3b. Decreasing Step Size (Inverse Square Root)

Implement a decreasing step size $\eta=c/\sqrt{t+1}$.
```python
def eta_sqrt(t,c=1e-3) -> float:
```

## 3c. Multi-Step Schedule

Implement a **piecewise-decaying step size** that drops by a factor `c` at predefined milestones:

$$\begin{aligned} \mathrm{eta\_multistep(t, [20,50], c=0.1, eta\_init=1)} = \begin{cases} 1, & t<20\\ 0.1 & 20\leq t<50\\ 0.01 & 50\leq t \end{cases} \end{aligned}$$

Implemented with the following interface:

```python
def eta_multistep(t, milestones=[20, 50], c=1e-4, eta_init=1e-3) -> float:
```

## 3d. Initialization

Repeat the above experiments (e.g., using `eta_const`) with different starting points $(u_0, v_0)$:

- $(-4, 0)$
- $(0, 0)$
- $(4, 0)$
- $(0, 4)$
- $(5, 5)$

For each initialization:

- Report the final point $(u_{100}, v_{100})$
- Report the final function value
- Optional (not graded): Plot the gradient descent trajectories of the different starting points.


# 4. Coordinate Descent

In optimization, we often aim to find the global minimum of a function by solving the gradient equation analytically. However, this becomes impractical in many real-world settings. Functions may be too complex, too high-dimensional, or lack closed-form solutions. This is where **coordinate descent** provides an alternative.

Coordinate descent updates one variable at a time, minimizing along each coordinate direction while holding the others fixed. This is particularly useful when:
- The function is differentiable but hard to minimize jointly,
- Partial updates are much easier to compute (analytically or numerically).

In this assignment, you'll apply coordinate descent to the following function:

$$
f(\mathbf{x}) = \exp(x_1 - 3x_2 + 3) + \exp(3x_2 - 2x_3 - 2) + \exp(2x_3 - x_1 + 2)
$$
with $x_1, x_2, x_3 \in \mathbb{R}$.

This function is fully differentiable, and its coordinate-wise argmins are analytically tractable. In part $a$, you’ll derive the updates for each coordinate individually. In part $b$, you’ll use these updates to implement a full coordinate descent loop, stepping through each coordinate iteratively and observing the convergence behavior.

## 4a. Analytically Computing Partial Gradients
Implement for each coordinate $ x_i $, $( i \in \{1,2,3\} )$ a function `argmin_xi(x)` that returns $\arg\min_{x_i} f(x)$, for each coordinate using the initial point $\mathbf{x}_{t_0} = (4, 3, 2) $:

- **$ \arg\min_{x_1} f({\mathbf{x}_{t_0}}) = $**
- **$ \arg\min_{x_2} f({\mathbf{x}_{t_0}}) = $**
- **$ \arg\min_{x_3} f({\mathbf{x}_{t_0}}) = $**

## 4b. Implementing The Coordinate Descent Loop
Implement a function `coordinate_descent(f, argmin, x_t0, max_iter=25)` that performs `max_iter` coordinate descent steps, where:

- `f` is the function to be minimized.
- `argmin` is an array of the `argmin_xi` functions for each coordinate.
- `x_t0` is the starting point (initialization).

At iteration $ t $, update all coordinates in order:
$$
x_t[i] = \text{argmin}[i](x_t)
$$

Using the initial point $\mathbf{x}_{t_0} = (1, 20, 5)$, run your coordinate descent implementation and answer the following:

- What are the final three coordinates (i.e. after the final step $t_n$)?
    - **$ \arg\min_{x_1} f({\mathbf{x}_{t_n}}) = $**
    - **$ \arg\min_{x_2} f({\mathbf{x}_{t_n}}) = $**
    - **$ \arg\min_{x_3} f({\mathbf{x}_{t_n}}) = $**
- What is the value the coordinate descent converges to?
  - $ f(\mathbf{x}_{t_n}) = $
- Optional (not graded): Use visualizations to validate your answer. Hint: A partial check you can perform is to see if you ended up in a local mimimum across dimensions.

# 5. Bias and Variance
Consider the true regression function:
$$
f^*(x) = \sigma(x) = \frac{1}{1 + e^{-x}}
$$
 
You might recognize this as the *sigmoid function*. Suppose we fit three regression models to different independent and identically distributed (i.i.d.) datasets sampled from the true regression function, $\mathcal{D}_1, \mathcal{D}_2, \mathcal{D}_3$, resulting in the following models:
$$
\begin{aligned}
f_{\mathcal{D}_1}(x) &= 2x + 0.4 \\
f_{\mathcal{D}_2}(x) &= x + 0.1 \\
f_{\mathcal{D}_3}(x) &= 3x + 0.7
\end{aligned}
$$

For $ x_0 = 0 $, compute:
1. The sample mean of the predictions $ f_{\mathcal{D}_i}(x_0) $, $ i=1,2,3 $.
2. The bias of the average predictor relative to $ f^*(x_0) $.
3. The variance of the predictors at $ x_0 $.

Then provide $bias^2$ and $variance$:

$bias^2 = $

$variance = $

In [3]:
import math

# 1. sample mean
def f_D1(x: float) -> float:
    return 2*x + 0.4

def f_D2(x: float) -> float:
    return x + 0.1

def f_D3(x: float) -> float:
    return 3*x + 0.7

x_0 = 0

sample_mean = round(np.mean([f_D1(x_0), f_D2(x_0), f_D3(x_0)]), 2)
print(f'sample mean: {sample_mean}')

# 2. bias
def true_regression(x: float) -> float:
    return 1/(1 + math.pow(math.e, -x))
true_value = true_regression(x_0)
bias_square = round((sample_mean - true_value)**2, 2)
print(f'bias^2: {bias_square}')

# 3. variance
deviations = [(f_D1(x_0) - sample_mean)**2, (f_D2(x_0) - sample_mean)**2, (f_D3(x_0) - sample_mean)**2]
variance = round(np.mean(deviations), 2)
print(f'variance: {variance}')


sample mean: 0.4
bias^2: 0.01
variance: 0.06


# 6. Polynomial Regression

In this exercise, you’ll explore regression, a technique that is used to predict continuous outputs, i.e. real numbers, based on multiple inputs. You will implement a full modeling pipeline from raw data to fitted models. The focus is on building an intuitive and practical understanding of how regression models behave as they grow in complexity.

You will use the California housing dataset. The accompanying Jupyter notebook includes instructions to load the data. Your task is to predict housing prices based on demographic and geographic features such as median income, housing age, and population density. You'll begin by exploring the dataset, which is crucial for any type of data based modeling, and then fit models with different types of tradeoffs.

## 6a. Exploring the Dataset

Before modeling, it's important to understand what kind of data you're working with. Even though you do not strictly need to do so to answer the question, consider using visualizations and summary statistics to explore relationships between variables and prices. Are any features clearly predictive? Are there outliers or skewed distributions? Consider how these factors might affect modeling later on. Building intuiting on the data you are modelling pretty much always pays off, especially when trying to identify and fix modelling problems and bugs.

Report the following:
- Number of samples in the dataset:
- Number of features in the dataset, excluding the target:

Optional task (not graded):
- Plot the relationship between the target (price) and the different features.

## 6b. Polynomial Feature Expansion

Linear regression can only capture straight-line relationships — but housing data often involves more complexity. To capture nonlinear patterns, you’ll expand the input features using a polynomial of various degrees, which includes squares and interaction terms between features. Use `PolynomialFeatures` from `sklearn.preprocessing` to construct this expanded design matrix.

However, polynomial features can lead to numerical instability, especially when the original features vary in scale. Large feature values produce large squared terms, which can cause issues during optimization. In practice, this results in warnings about ill-conditioned matrices. To avoid this, you’ll standardize the original data using `StandardScaler` before generating polynomial features.

Once the transformation is complete, report the shape of the polynomial design matrix, after expansion with a polynomial of degree 2:

**Do not include bias when constructing the features using PolynomialFeatures**.

These observations should help you develop intuition for the cost of model complexity.

## 6c. Fitting A Regression
With your polynomial design matrix in hand, you'll now compute the regression model that minimizes the residual sum of squares (RSS) and compare it to the performance of a linear model. 

**Use the data set split in the accompanying notebook to answer the following questions**. You will have a training and validation dataset. Only use the training dataset to fit the models.  

After solving for the regression parameters, report the following parameters for each of two models, and report the mean squared error (MSE) on the validation set:

Linear model:
- $ \beta_{\text{MedInc}} $
- $ \beta_{\text{AveBedrms}} $
- $ \beta_{\text{HouseAge}} $
- $MSE_{val} = $ 

Polynomial of degree 2:
- $ \beta_{\text{MedInc}} $
- $ \beta_{\text{MedInc} \cdot \text{AveBedrms}} $
- $ \beta_{\text{HouseAge} \cdot \text{AveBedrms}} $
- $MSE = $

Optional (not graded): Plot data sample of your trained models on top of the data to get a sense of the model fit for different target-feature combinations.

In [4]:
from sklearn.datasets import fetch_california_housing 
import pandas as pd
from sklearn.preprocessing import StandardScaler,PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Load the dataset
housing = fetch_california_housing(as_frame=True)

# The data is in a pandas DataFrame
df = housing.frame

# View the first few rows


features=df.drop(columns=["MedHouseVal"])
target=df['MedHouseVal']

# No. of samples in the data set=20640
# No. of features= 8 numeric, predictive attribes

def standardscaler(X):
    scaler=StandardScaler()
    return scaler.fit_transform(X)

scaled_features=standardscaler(features)


X_train, X_test, y_train, y_test = train_test_split(scaled_features, target, test_size=0.2, random_state=42)

#Linear regression

lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

linear_coefs = pd.Series(lin_reg.coef_, index=features.columns) #coefficents for linear regression model



y_pred_lin = lin_reg.predict(X_test)
mse_lin = mean_squared_error(y_test, y_pred_lin) #mean square error for linear regression 


#Polynomial regression degree 2


poly=PolynomialFeatures(2, include_bias=False)
X_poly_train=poly.fit_transform(X_train)
X_poly_test=poly.fit_transform(X_test)

poly_reg=LinearRegression()
poly_reg.fit(X_poly_train,y_train)

poly_coefs = pd.Series(poly_reg.coef_, index=poly.get_feature_names_out(features.columns))#coefficents for polynomial regression model


y_pred_poly = poly_reg.predict(X_poly_test)
mse_ploy=mean_squared_error(y_test,y_pred_poly) #mean square error for polynomial regression 

print(poly_coefs)

#The results are in the markdown cell below

MedInc                  0.934537
HouseAge                0.132364
AveRooms               -0.404129
AveBedrms               0.581859
Population              0.039435
AveOccup               -1.599901
Latitude               -1.272053
Longitude              -1.166391
MedInc^2               -0.111700
MedInc HouseAge         0.037707
MedInc AveRooms         0.185882
MedInc AveBedrms       -0.131132
MedInc Population       0.110720
MedInc AveOccup        -0.088451
MedInc Latitude        -0.665435
MedInc Longitude       -0.584181
HouseAge^2              0.033202
HouseAge AveRooms      -0.016815
HouseAge AveBedrms      0.057186
HouseAge Population     0.035832
HouseAge AveOccup      -0.249630
HouseAge Latitude      -0.276318
HouseAge Longitude     -0.252209
AveRooms^2              0.064875
AveRooms AveBedrms     -0.124241
AveRooms Population    -0.159720
AveRooms AveOccup       0.537247
AveRooms Latitude       0.563111
AveRooms Longitude      0.495969
AveBedrms^2             0.059291
AveBedrms 

**Results**

Linear model:
- $ \beta_{\text{MedInc}}=0.934537 $
- $ \beta_{\text{AveBedrms}}=0.581859 $
- $ \beta_{\text{HouseAge}}=0.122382 $
- $MSE_{val} =0.5558915986952442 $ 


Polynomial of degree 2:
- $ \beta_{\text{MedInc}}=0.934537 $
- $ \beta_{\text{MedInc} \cdot \text{AveBedrms}}=-0.131132 $
- $ \beta_{\text{HouseAge} \cdot \text{AveBedrms}}=-0.124241 $
- $MSE =0.46430152383013235 $


# 7. Regularization And Cross Validation

Building on your polynomial regression models from Exercise 6, you'll now explore two critical concepts in machine learning: regularization to prevent overfitting, and cross-validation to obtain more reliable performance estimates. These techniques address fundamental challenges that arise when working with complex models on real data.
In Exercise 6, you likely observed that the polynomial model achieved lower training error than the linear model, but this doesn't necessarily mean it will generalize better to new data. Exercise 7 introduces tools to address this challenge systematically.

## 7a. Regularization

To counter overfitting and improve stability, you’ll re-fit the models using **Ridge regression**, which penalizes large weights via an L2 penalty. The specific objective here is a _modified version_ of the standard ridge regression objective function, with $n$ indicating the number of data points used to fit the model:

$$
\min_{\beta} \frac{1}{n} \| y - X\beta \|^2 + \lambda \| \beta \|^2
$$

Implement ridge regression using the same linear and polynomial design matrix, and set $ \lambda = 0.001 $. As before, examine how the regularization changes the learned parameters and the mean squared error (MSE).

Report the same three coefficients:

Linear model:
- $ \beta_{\text{MedInc}} $
- $ \beta_{\text{AveBedrms}} $
- $ \beta_{\text{HouseAge}} $
- $MSE_{val} = $ 

Polynomial of degree 2:
- $ \beta_{\text{MedInc}} =$
- $ \beta_{\text{MedInc} \cdot \text{AveBedrms}} =$
- $ \beta_{\text{HouseAge} \cdot \text{AveBedrms}} =$
- $MSE_{val} = $

## 7b. Cross-Validation

Cross-validation provides a more robust estimate of model performance than a single train-validation split, at the cost of significantly more computation. By systematically training and evaluating on different subsets of the data, you can better understand how well your models generalize and reduce the risk of making decisions based on a particular "lucky" or "unlucky" split.

**Use the 5-fold cross-validation splits provided in the setup notebook to answer the following question.** For each fold, you'll train your models on four folds and evaluate on the fifth, repeating this process for all five folds. This gives you five performance estimates that you can aggregate to get a more reliable assessment of model quality.

You'll evaluate both the linear model and the polynomial degree 2 model using the same Ridge regression approach from 7a. Remember to apply the same preprocessing pipeline within each fold: standardize the training folds, apply polynomial expansion if needed, then fit the ridge regression model.

**Important**: Each fold should be treated as an independent experiment. This means you should standardize features using only the training folds for that iteration, not the entire dataset. This prevents data leakage and ensures your cross-validation estimates are unbiased.

After completing cross-validation, also evaluate both models on the held-out validation set (`X_val`, `y_val`) to assess final performance on truly unseen data.

Report the following metrics:

**Linear model (Ridge):**
- Mean Cross-Validation MSE = 
- Standard deviation of Cross-Validation MSE = 
- Final validation MSE = 

**Polynomial degree 2 model (Ridge):**
- Mean Cross-Validation MSE = 
- Standard deviation of Cross-Validation MSE = 
- Final validation MSE = 

**Optional task (not graded):** How do the cross-validation estimates compare to the final validation performance? What does this tell you about the reliability of your model selection process? How do the measures compare to a single train-test split? Did regularization help performance?

In [5]:
from sklearn.datasets import fetch_california_housing
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

# Load the dataset
housing = fetch_california_housing(as_frame=True)

# The data is in a pandas DataFrame
df = housing.frame

# Splitting the dataset
from sklearn.model_selection import train_test_split

# Let's separate features and target for clarity
features = df.drop(columns=["MedHouseVal"])
target = df["MedHouseVal"]

# Standardize the features
# Done is exercise 6
scaler=StandardScaler()
scaled_features=scaler.fit_transform(features)

# Split the dataset (80% train, 20% validation)
X_train, X_val, y_train, y_val = train_test_split(scaled_features, target, test_size=0.2, random_state=42)

# Linear Ridge

ridge_alpha = 0.001 * len(X_train)  # ?????
lin_ridge = Ridge(alpha=ridge_alpha, fit_intercept=True)
lin_ridge.fit(X_train, y_train)

y_val_pred_lin = lin_ridge.predict(X_val)
mse_val_lin = mean_squared_error(y_val, y_val_pred_lin)

lin_coefs = pd.Series(lin_ridge.coef_, index=features.columns)

beta_MedInc_lin = lin_coefs.get("MedInc", np.nan)
beta_AveBedrms_lin = lin_coefs.get("AveBedrms", np.nan)
beta_HouseAge_lin = lin_coefs.get("HouseAge", np.nan)

# Polynomial Ridge
poly = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly.fit_transform(X_train)
X_val_poly = poly.transform(X_val)

poly_feat_names = poly.get_feature_names_out(features.columns)

poly_ridge = Ridge(alpha=ridge_alpha, fit_intercept=True)
poly_ridge.fit(X_train_poly, y_train)

y_val_pred_poly = poly_ridge.predict(X_val_poly)
mse_val_poly = mean_squared_error(y_val, y_val_pred_poly)

poly_coefs = pd.Series(poly_ridge.coef_, index=poly_feat_names)

beta_MedInc_poly = poly_coefs.get("MedInc", np.nan)
beta_MedInc_AveBedrms_poly = poly_coefs.get("MedInc AveBedrms", np.nan)
beta_HouseAge_AveBedrms_poly = poly_coefs.get("HouseAge AveBedrms", np.nan)

# Report (7a)
print("7a. Ridge (lambda=0.001; sklearn alpha=0.0005)")
print("\nLinear model:")
print("β_MedInc =", beta_MedInc_lin)
print("β_AveBedrms =", beta_AveBedrms_lin)
print("β_HouseAge =", beta_HouseAge_lin)
print("MSE_val =", mse_val_lin)

print("\nPolynomial degree 2 model:")
print("β_MedInc =", beta_MedInc_poly)
print("β_MedInc·AveBedrms =", beta_MedInc_AveBedrms_poly)
print("β_HouseAge·AveBedrms =", beta_HouseAge_AveBedrms_poly)
print("MSE_val =", mse_val_poly)

print("\n[Shape of degree-2 design matrix (train)]:", X_train_poly.shape)



7a. Ridge (lambda=0.001; sklearn alpha=0.0005)

Linear model:
β_MedInc = 0.8511882602212492
β_AveBedrms = 0.3659930686182285
β_HouseAge = 0.12364455466453653
MSE_val = 0.555257574854269

Polynomial degree 2 model:
β_MedInc = 0.9405701833600948
β_MedInc·AveBedrms = -0.12524810942955888
β_HouseAge·AveBedrms = 0.0671055170285
MSE_val = 0.44329120116660675

[Shape of degree-2 design matrix (train)]: (16512, 44)


In [6]:
# Load the dataset
housing = fetch_california_housing(as_frame=True)
df = housing.frame

# Separate features and target
features = df.drop(columns=["MedHouseVal"])
target = df["MedHouseVal"]

# Split the dataset (80% train, 20% validation) - using raw features
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(features, target, test_size=0.2, random_state=42)

# Cross-validation splits for exercise 7b
from sklearn.model_selection import KFold
cv_splits = list(KFold(n_splits=5, shuffle=True, random_state=42).split(X_train))

print(f"Dataset loaded: {len(df)} total samples")
print(f"Training set: {len(X_train)} samples")
print(f"Validation set: {len(X_val)} samples")
print(f"Created {len(cv_splits)} folds for cross-validation")
print(f"Each fold uses ~{len(X_train)//5} samples for validation")

def cv_ridge_linear(X_train, y_train, splits, alpha):
    mses = []
    for tr_idx, va_idx in splits:
        X_tr = X_train.iloc[tr_idx]
        y_tr = y_train.iloc[tr_idx]
        X_va = X_train.iloc[va_idx]
        y_va = y_train.iloc[va_idx]

        # Scale within fold
        scaler = StandardScaler()
        X_tr_scaled = pd.DataFrame(
            scaler.fit_transform(X_tr),
            columns=X_tr.columns,
            index=X_tr.index
        )
        X_va_scaled = pd.DataFrame(
            scaler.transform(X_va),
            columns=X_va.columns,
            index=X_va.index
        )

        # Fit Ridge
        model = Ridge(alpha=alpha, fit_intercept=True, random_state=42)
        model.fit(X_tr_scaled, y_tr)
        y_hat = model.predict(X_va_scaled)
        mses.append(mean_squared_error(y_va, y_hat))
    return np.array(mses)

def cv_ridge_poly2(X_train, y_train, splits, alpha):
    mses = []
    for tr_idx, va_idx in splits:
        X_tr = X_train.iloc[tr_idx]
        y_tr = y_train.iloc[tr_idx]
        X_va = X_train.iloc[va_idx]
        y_va = y_train.iloc[va_idx]

        # Scale within fold
        scaler = StandardScaler()
        X_tr_scaled = pd.DataFrame(
            scaler.fit_transform(X_tr),
            columns=X_tr.columns,
            index=X_tr.index
        )
        X_va_scaled = pd.DataFrame(
            scaler.transform(X_va),
            columns=X_va.columns,
            index=X_va.index
        )

        # Polynomial features within fold (fit on fold-train only)
        poly = PolynomialFeatures(degree=2, include_bias=False)
        X_tr_poly = poly.fit_transform(X_tr_scaled)
        X_va_poly = poly.transform(X_va_scaled)

        # Fit Ridge
        model = Ridge(alpha=alpha, fit_intercept=True, random_state=42)
        model.fit(X_tr_poly, y_tr)
        y_hat = model.predict(X_va_poly)
        mses.append(mean_squared_error(y_va, y_hat))
    return np.array(mses)

ridge_alpha = 0.001 * len(X_train)  # ?????

cv_mses_lin = cv_ridge_linear(X_train, y_train, cv_splits, alpha=ridge_alpha) #?
cv_mses_poly = cv_ridge_poly2(X_train, y_train, cv_splits, alpha=ridge_alpha) #?

print("7b — Linear Ridge:")
print("Mean CV MSE =", cv_mses_lin.mean())
print("Std CV MSE  =", cv_mses_lin.std())

print("\n7b — Polynomial (degree 2) Ridge:")
print("Mean CV MSE =", cv_mses_poly.mean())
print("Std CV MSE  =", cv_mses_poly.std())

# Final validation
# Linear (refit on full train with proper scaling)
scaler_final = StandardScaler()
X_train_scaled_final = scaler_final.fit_transform(X_train)
X_val_scaled_final = scaler_final.transform(X_val)

ridge_lin_final = Ridge(alpha=ridge_alpha, fit_intercept=True, random_state=42)
ridge_lin_final.fit(X_train_scaled_final, y_train)
y_val_pred_lin = ridge_lin_final.predict(X_val_scaled_final)
final_val_mse_lin = mean_squared_error(y_val, y_val_pred_lin)

# Polynomial (refit on full train with fold-like preprocessing)
poly_final = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly_final = poly_final.fit_transform(X_train_scaled_final)
X_val_poly_final = poly_final.transform(X_val_scaled_final)

ridge_poly_final = Ridge(alpha=ridge_alpha, fit_intercept=True, random_state=42)
ridge_poly_final.fit(X_train_poly_final, y_train)
y_val_pred_poly = ridge_poly_final.predict(X_val_poly_final)
final_val_mse_poly = mean_squared_error(y_val, y_val_pred_poly)

print("\nFinal validation MSE (held-out):")
print("Linear Ridge:    ", final_val_mse_lin)
print("Poly (deg=2) Ridge:", final_val_mse_poly)


Dataset loaded: 20640 total samples
Training set: 16512 samples
Validation set: 4128 samples
Created 5 folds for cross-validation
Each fold uses ~3302 samples for validation
7b — Linear Ridge:
Mean CV MSE = 0.5193529397563056
Std CV MSE  = 0.019909320703632806

7b — Polynomial (degree 2) Ridge:
Mean CV MSE = 0.6682719487221405
Std CV MSE  = 0.3568754263960067

Final validation MSE (held-out):
Linear Ridge:     0.5553151596305632
Poly (deg=2) Ridge: 0.44515578615994433


# 8. California Housing Dataset

This open question will test your understanding of the entire modeling process. Use the California housing dataset (see the setup notebook).

## 8a. Feature Selection

What are the candidate features of this dataset that could be removed according to variance thresholding or correlation-based feature selection? Explain how you derive the candidates by means of plots or tables and choose at least three features that are candidates for removal.

## 8b. Cross Validation

Implement a function that returns the regression model for a given design matrix and target vector. Fit and evaluate four regression models (with affine basis functions) using (I) all features and (II-IV) all but one of the three candidate features from the previous part. Use 5-fold cross-validation to evaluate your regression models. Add a table in the report with the cross-validated scores for each of your regression models.

## 8c. Interpretation

Interpret your results. Would you recommend removing one of the candidate features? What would you infer from the cross-validated scores?

Justify your analysis and discuss possible benefits and drawbacks when removing one of the features vs. keeping all features.